### PR TITLE
Ccip-4158 - Initial Add Chain changeset

### PR DESCRIPTION
**Pull request by AnieeG**
- It addresses the InitialAddChain part in ticket [CCIP-4158](https://smartcontract-it.atlassian.net/browse/CCIP-4158) 
    - Creates a changeset to AddChainConfig + AddDON (candidate->primary promotion i.e. init) on the home chain, SetOCR3Config on the remote chain
    
- Moves USDC contract deployment to DeployPrerequisites as these contracts are separate from ccip.
 
- Creates `ChangesetApplication.Apply ` so that multiple changesets can be applied from test as it would be followed in real deployment.
- Moves SaveExistingChangeset to common/changeset so that it can be used by all
- Makes all contract deployment across different chains parallel.
- Creates separate go routine for funding of nodes in different chains
- Updates few function to return error instead of relying on require.NoError

[CCIP-4158]: https://smartcontract-it.atlassian.net/browse/CCIP-4158?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: deployment/ccip/changeset/deploy.go [hunk 1]
```diff
@@ -67,25 +63,35 @@ var (
...
+		})
 	}
-	return nil
+	return deployGrp.Wait()
 }
 
 // DeployPrerequisiteContracts deploys the contracts that can be ported from previous CCIP version to the new one.
 // This is only required for staging and test environments where the contracts are not already deployed.
-func DeployPrerequisiteContracts(e deployment.Environment, ab deployment.AddressBook, state CCIPOnChainState, chain deployment.Chain) error {
+func DeployPrerequisiteContracts(e deployment.Environment, ab deployment.AddressBook, state CCIPOnChainState, chain deployment.Chain, usdcEnabled bool) error {
```
Comment (makramkd) [inline deployment/ccip/changeset/deploy.go]: Can we make this function take a variadic or something at the end? I assume I can do something similar to this in the multicall3 PR

```
type Options struct {
  USDCEnabled bool
  Multicall3Enabled bool
}

type DeployOpt func(o *Options)

func WithUSDC(enabled bool) DeployOpt {
  return func(o *Options) {
    o.USDCEnabled = enabled
  }
}

func WithMulticall3(enabled bool) DeployOpt {
  return func(o *Options) {
    o.Multicall3Enabled = enabled
  }
}

func DeployPrerequisiteContracts(e deployment.Environment, ab deployment.AddressBook, state CCIPOnChainState, chain deployment.Chain, deployOpts ...DeployOpt) {
  options := &Options{}
  for _, deployOpt := range deployOpts {
    deployOpt(options)
  }

  // use options.USDCEnabled
}
```

File: deployment/ccip/changeset/test_helpers.go [hunk 2]
```diff
@@ -242,15 +244,10 @@ func mockAttestationResponse() *httptest.Server {
 	return server
 }
 
-func NewMemoryEnvironmentWithJobsAndContracts(t *testing.T, lggr logger.Logger, numChains int, numNodes int) DeployedEnv {
+func NewMemoryEnvironmentWithJobsAndContracts(t *testing.T, lggr logger.Logger, numChains int, numNodes int, isUSDC bool) DeployedEnv {
```
Comment (makramkd) [inline deployment/ccip/changeset/test_helpers.go]: I think this would benefit greatly from using the options pattern I mentioned above, everything after `lggr logger.Logger` can be an option

File: deployment/common/changeset/apply.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	"github.com/smartcontractkit/ccip-owner-contracts/pkg/gethwrappers"
+	jobv1 "github.com/smartcontractkit/chainlink-protos/job-distributor/v1/job"
+
+	"github.com/smartcontractkit/chainlink/deployment"
+)
+
+type ChangesetApplication struct {
+	Changeset func(e deployment.Environment, config any) (deployment.ChangesetOutput, error)
+	Config    any
+}
```
Comment (makramkd) [inline deployment/common/changeset/apply.go]: Can this use generics and be typesafe instead of `any`?

```golang
type ChangesetApplication[T any] struct {
	Changeset func(e deployment.Environment, config T) (deployment.ChangesetOutput, error)
	Config    T
}
```

Comment (makramkd) [reply deployment/common/changeset/apply.go]: Also please add a comment for what this type's purpose is

File: deployment/common/changeset/apply.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+			}
+			return fn(e, config.(C))
+		}
+
+		return fn(e, zeroC)
+	}
+}
+
+// ApplyChangesets applies the changeset applications to the environment and returns the updated environment.
+func ApplyChangesets(ctx context.Context, e deployment.Environment, timelocksPerChain map[uint64]*gethwrappers.RBACTimelock, changesetApplications []ChangesetApplication) (deployment.Environment, error) {
```
Comment (makramkd) [inline deployment/common/changeset/apply.go]: I see, ChangesetApplication here can have a bunch of different config types, so can't be generic

File: deployment/common/changeset/apply.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+				signed, err := SignProposal(e, &prop)
+				if err != nil {
+					return deployment.Environment{}, fmt.Errorf("failed to sign proposal: %w", err)
+				}
+				for _, sel := range chains.ToSlice() {
+					timelock, ok := timelocksPerChain[sel]
+					if !ok || timelock == nil {
+						return deployment.Environment{}, fmt.Errorf("timelock not found for chain %d", sel)
+					}
+					err := ExecuteProposal(e, signed, timelock, sel)
```
Comment (makramkd) [inline deployment/common/changeset/apply.go]: Is this func expected to be used only in tests? Signs + executes the proposal seems to suggest that, but not 100% sure

File: integration-tests/ccip-tests/testsetups/test_helpers.go [hunk 6]
```diff
@@ -74,16 +77,16 @@ func (d DeployedLocalDevEnvironment) RestartChainlinkNodes(t *testing.T) error {
 	return errGrp.Wait()
 }
 
-func NewLocalDevEnvironmentWithDefaultPrice(
-	t *testing.T,
-	lggr logger.Logger) (changeset.DeployedEnv, *test_env.CLClusterTestEnv, testconfig.TestConfig) {
-	return NewLocalDevEnvironment(t, lggr, changeset.MockLinkPrice, changeset.MockWethPrice)
+func NewLocalDevEnvironmentWithDefaultPrice(t *testing.T, lggr logger.Logger, isUSDC bool) (changeset.DeployedEnv, *test_env.CLClusterTestEnv, tc.TestConfig) {
```
Comment (makramkd) [inline integration-tests/ccip-tests/testsetups/test_helpers.go]: Ditto here and below re: options

File: integration-tests/ccip-tests/testsetups/test_helpers.go [hunk 7]
```diff
@@ -109,95 +112,131 @@ func NewLocalDevEnvironment(
...
+		},
+		{
+			Changeset: commonchangeset.WrapChangeSet(changeset.DeployChainContracts),
+			Config: changeset.DeployChainContractsConfig{
+				ChainSelectors:    allChains,
+				HomeChainSelector: homeChainSel,
+			},
+		},
+	}
+	ctx = testcontext.Get(t)
```
Comment (makramkd) [inline integration-tests/ccip-tests/testsetups/test_helpers.go]: This seems unnecessary

File: integration-tests/ccip-tests/testsetups/test_helpers.go [hunk 8]
```diff
@@ -109,95 +112,131 @@ func NewLocalDevEnvironment(
...
+	for _, c := range env.AllChainSelectors() {
+		mcmsCfg[c] = mcmsCfgPerChain
 	}
-	output, err = commonchangeset.DeployMCMSWithTimelock(*e, mcmsCfg)
-	require.NoError(t, err)
-	require.NoError(t, e.ExistingAddresses.Merge(output.AddressBook))
-
-	state, err := changeset.LoadOnchainState(*e)
+	// Need to deploy prerequisites first so that we can for the CCIP contracts.
+	cs01 := []commonchangeset.ChangesetApplication{
```
Comment (makramkd) [inline integration-tests/ccip-tests/testsetups/test_helpers.go]:
```suggestion
	prereqChangesets := []commonchangeset.ChangesetApplication{
```

File: integration-tests/ccip-tests/testsetups/test_helpers.go [hunk 9]
```diff
@@ -109,95 +112,131 @@ func NewLocalDevEnvironment(
...
-		ChainsToDeploy: e.AllChainSelectors(),
-		TokenConfig:    tokenConfig,
-		OCRSecrets:     deployment.XXXGenerateTestOCRSecrets(),
-		USDCConfig: changeset.USDCConfig{
-			Enabled: true,
-			USDCAttestationConfig: changeset.USDCAttestationConfig{
-				API:         endpoint,
-				APITimeout:  commonconfig.MustNewDuration(time.Second),
-				APIInterval: commonconfig.MustNewDuration(500 * time.Millisecond),
+	USDCCCTPConfig := make(map[cciptypes.ChainSelector]pluginconfig.USDCCCTPTokenConfig)
```
Comment (makramkd) [inline integration-tests/ccip-tests/testsetups/test_helpers.go]:
```suggestion
	usdcCCTPConfig := make(map[cciptypes.ChainSelector]pluginconfig.USDCCCTPTokenConfig)
```

File: integration-tests/ccip-tests/testsetups/test_helpers.go [hunk 10]
```diff
@@ -109,95 +112,131 @@ func NewLocalDevEnvironment(
...
+		endpoint = testEnv.MockAdapter.InternalEndpoint
+		usdcAttestationCfg = changeset.USDCAttestationConfig{
+			API:         endpoint,
+			APITimeout:  commonconfig.MustNewDuration(time.Second),
+			APIInterval: commonconfig.MustNewDuration(500 * time.Millisecond),
+		}
+	}
+
+	// Deploy second set of changesets to deploy and configure the CCIP contracts.
+	cs02 := []commonchangeset.ChangesetApplication{
```
Comment (makramkd) [inline integration-tests/ccip-tests/testsetups/test_helpers.go]:
```suggestion
	deployCCIPChangesets := []commonchangeset.ChangesetApplication{
```

File: deployment/common/changeset/apply.go (hunk 5 above)
Comment (connorwstein) [reply deployment/common/changeset/apply.go]: yeah it is. Lets name the file test_helpers.go to be clear

File: deployment/common/changeset/mcms_test_helpers.go [hunk 11]
```diff
@@ -42,62 +43,81 @@ func SingleGroupMCMS(t *testing.T) config.Config {
 	return *c
 }
 
-func SignProposal(t *testing.T, env deployment.Environment, proposal *timelock.MCMSWithTimelockProposal) *mcms.Executor {
```
Comment (connorwstein) [inline deployment/common/changeset/mcms_test_helpers.go]: why the change here? its kind of nice that the testing arg makes it clear its for test only

File: deployment/common/changeset/save_existing.go [hunk 12]
```diff
@@ -42,6 +42,8 @@ func (cfg ExistingContractsConfig) Validate() error {
 	return nil
 }
 
+// SaveExistingContracts saves the existing contracts to the address book.
+// Caller should update the environment's address book with the returned addresses.
 func SaveExistingContracts(env deployment.Environment, cfg ExistingContractsConfig) (deployment.ChangesetOutput, error) {
```
Comment (connorwstein) [inline deployment/common/changeset/save_existing.go:47]: can we move the test here too and modify it to be abstract? thats more like a unit test which should be as close as possible to the code its testing

File: deployment/ccip/changeset/test_helpers.go (hunk 2 above)
Comment (connorwstein) [reply deployment/ccip/changeset/test_helpers.go]: in this case maybe just a TestConfig struct would be easier? Its more like regular configuration so option funcs might be overkill

File: deployment/common/changeset/apply.go [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+	mapset "github.com/deckarep/golang-set/v2"
+	"github.com/smartcontractkit/ccip-owner-contracts/pkg/gethwrappers"
+	jobv1 "github.com/smartcontractkit/chainlink-protos/job-distributor/v1/job"
+
+	"github.com/smartcontractkit/chainlink/deployment"
+)
+
+type ChangesetApplication struct {
+	Changeset func(e deployment.Environment, config any) (deployment.ChangesetOutput, error)
```
Comment (connorwstein) [inline deployment/common/changeset/apply.go]: you can do `deployment.ChangeSet[any]` here no need to redefine

File: deployment/ccip/changeset/test_helpers.go [hunk 14]
```diff
@@ -262,37 +259,105 @@ func NewMemoryEnvironmentWithJobsAndContracts(t *testing.T, lggr logger.Logger,
...
+					CCTPTokenConfig:       USDCCCTPConfig,
+				},
+			},
+		},
+		{
+			Changeset: commonchangeset.WrapChangeSet(CCIPCapabilityJobspec),
+		},
+	}
+
+	e.Env, err = commonchangeset.ApplyChangesets(context.Background(), e.Env, timelocksPerChain, cs02)
```
Comment (connorwstein) [inline deployment/ccip/changeset/test_helpers.go]: could we inline the list of changeset here? I think it'd be more readable if it was like ApplyChangesets(ctx, e.Env, []common.Changesets{.....})

File: deployment/ccip/changeset/initial_deploy_test.go [hunk 15]
```diff
@@ -21,61 +15,11 @@ import (
 
 func TestInitialDeploy(t *testing.T) {
```
Comment (connorwstein) [inline deployment/ccip/changeset/initial_deploy_test.go:16]: should remove this test and have an initial_add_chain_test.go which covers the add chain functionality

File: deployment/ccip/changeset/deploy.go [hunk 16]
```diff
@@ -67,25 +63,35 @@ var (
 	USDCTokenPool       deployment.ContractType = "USDCTokenPool"
 )
 
-func DeployPrerequisiteChainContracts(e deployment.Environment, ab deployment.AddressBook, selectors []uint64) error {
+func DeployPrerequisiteChainContracts(e deployment.Environment, ab deployment.AddressBook, selectors []uint64, usdcEnabledChains []uint64) error {
```
Comment (connorwstein) [inline deployment/ccip/changeset/deploy.go]: lets make this function private to enforce that external packages (integration-tests, chainlink-deployment etc.) only use the changeset interface

File: deployment/ccip/changeset/deploy.go [hunk 17]
```diff
@@ -265,43 +271,31 @@ func DeployPrerequisiteContracts(e deployment.Environment, ab deployment.Address
...
-// It then deploys the rest of the CCIP chain contracts to the selected chains
-// registers the nodes with the capability registry and creates a DON for
-// each new chain. TODO: Might be better to break this down a bit?
-func DeployCCIPContracts(
+// ConfigureChain assumes the all the Home chain contracts and CCIP contracts are deployed
+// It does -
+// 1. AddChainConfig for each chain in CCIPHome
+// 2. Registers the nodes with the capability registry
+// 3. SetOCR3Config on the remote chain
+func ConfigureChain(
```
Comment (connorwstein) [inline deployment/ccip/changeset/deploy.go]: lets make it private

File: deployment/ccip/changeset/deploy.go [hunk 18]
```diff
@@ -422,6 +380,113 @@ func DeployCCIPContracts(
...
+	return m
+}
+
+type USDCAttestationConfig struct {
+	API         string
+	APITimeout  *commonconfig.Duration
+	APIInterval *commonconfig.Duration
+}
+
+type DeployCCIPContractConfig struct {
```
Comment (connorwstein) [inline deployment/ccip/changeset/deploy.go]: since this is top level changeset input config, lets put it in the changeset file itself

File: deployment/ccip/changeset/initial_add_chain.go [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+)
+
+var _ deployment.ChangeSet[DeployCCIPContractConfig] = InitialAddChain
+
+// InitialAddChain enables new chains as destination for CCIP
+// It performs the following steps:
+// - AddChainConfig + AddDON (candidate->primary promotion i.e. init) on the home chain
+// - SetOCR3Config on the remote chain
+// ConfigureChain assumes that the home chain is already enabled and all CCIP contracts are already deployed.
+func InitialAddChain(env deployment.Environment, c DeployCCIPContractConfig) (deployment.ChangesetOutput, error) {
```
Comment (connorwstein) [inline deployment/ccip/changeset/initial_add_chain.go]: should be plural as its adding multiple? Maybe just ConfigureNewChains

File: deployment/ccip/changeset/initial_add_chain.go [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+
+import (
+	"fmt"
+
+	"github.com/smartcontractkit/ccip-owner-contracts/pkg/proposal/timelock"
+
+	"github.com/smartcontractkit/chainlink/deployment"
+)
+
+var _ deployment.ChangeSet[DeployCCIPContractConfig] = InitialAddChain
```
Comment (connorwstein) [inline deployment/ccip/changeset/initial_add_chain.go]: definitely should rename this config struct - NewChainsConfig? Also lets keep the definition of the config here

File: deployment/ccip/changeset/deploy.go [hunk 21]
```diff
@@ -422,6 +380,113 @@ func DeployCCIPContracts(
...
+	HomeChainSel   uint64
+	FeedChainSel   uint64
+	ChainsToDeploy []uint64
+	TokenConfig    TokenConfig
```
Comment (connorwstein) [inline deployment/ccip/changeset/deploy.go]: hmm doesn't the token config vary by chain @makramkd ?

File: deployment/ccip/changeset/deploy.go [hunk 22]
```diff
@@ -422,6 +380,113 @@ func DeployCCIPContracts(
...
+	USDCConfig     USDCConfig
+	// For setting OCR configuration
+	OCRSecrets deployment.OCRSecrets
+}
```
Comment (connorwstein) [inline deployment/ccip/changeset/deploy.go]: we need to parameterize the chain configuration values as well those are just hard coded right now https://github.com/smartcontractkit/chainlink/blob/b375b3ca23dd4ff5c4c4524e9773e72ea397b58b/deployment/ccip/changeset/deploy_home_chain.go#L294

Comment (connorwstein) [reply deployment/ccip/changeset/deploy.go]: also I think the OCR timing parameters need to be parameterized and could vary by chain

**Review by connorwstein — CHANGES_REQUESTED** [review]

File: deployment/ccip/changeset/deploy.go (hunk 22 above)
Comment (AnieeG) [reply deployment/ccip/changeset/deploy.go]: added ticket ccip-4320

File: deployment/ccip/changeset/initial_deploy_test.go (hunk 15 above)
Comment (AnieeG) [reply deployment/ccip/changeset/initial_deploy_test.go:16]: I did not remove it because this is the only test chainlink-ccip runs

File: deployment/common/changeset/save_existing.go (hunk 12 above)
Comment (AnieeG) [reply deployment/common/changeset/save_existing.go:47]: I added a test here but kept the ccip specific test too

File: deployment/ccip/changeset/deploy.go [hunk 23]
```diff
@@ -67,25 +61,62 @@ var (
...
 
-func DeployPrerequisiteChainContracts(e deployment.Environment, ab deployment.AddressBook, selectors []uint64) error {
+type DeployPrerequisiteContractsOpts struct {
+	USDCEnabledChains []uint64
+	Multicall3Enabled bool
+}
+
+type PrerequisiteOpt func(o *DeployPrerequisiteContractsOpts)
+
+func WithUSDCChains(chains []uint64) PrerequisiteOpt {
```
Comment (connorwstein) [inline deployment/ccip/changeset/deploy.go:71]: these still don't look very useful to me, whats wrong with just using the struct directly?

File: deployment/ccip/changeset/deploy.go [hunk 24]
```diff
@@ -422,6 +405,38 @@ func DeployCCIPContracts(
...
+// DeployCCIPContracts assumes the following contracts are deployed:
+// - Capability registry
+// - CCIP home
+// - RMN home
+// - Fee tokens on all chains.
+// and present in ExistingAddressBook.
+// It then deploys the rest of the CCIP chain contracts to the selected chains
+// registers the nodes with the capability registry and creates a DON for
+// each new chain.
+func DeployCCIPContracts(
```
Comment (connorwstein) [inline deployment/ccip/changeset/deploy.go]: can be private

File: deployment/ccip/changeset/deploy.go [hunk 25]
```diff
@@ -422,6 +405,38 @@ func DeployCCIPContracts(
...
+	err = configureChain(e, c)
+	if err != nil {
+		e.Logger.Errorw("Failed to add chain", "err", err)
+		return err
+	}
+
+	return nil
+}
+
 func DeployChainContractsForChains(
```
Comment (connorwstein) [inline deployment/ccip/changeset/deploy.go]: can be private

**Review by makramkd — DISMISSED** [review]

File: deployment/ccip/changeset/deploy.go [hunk 26]
```diff
@@ -473,11 +491,19 @@ func DeployChainContractsForChains(
...
 			return fmt.Errorf("fee tokens not found for chain %d", chainSel)
 		}
-		err := deployChainContracts(e, chain, ab, rmnHome)
-		if err != nil {
-			e.Logger.Errorw("Failed to deploy chain contracts", "chain", chainSel, "err", err)
-			return fmt.Errorf("failed to deploy chain contracts for chain %d: %w", chainSel, err)
-		}
+		deployGrp.Go(
+			func() error {
+				err := deployChainContracts(e, chain, ab, rmnHome)
```
Comment (connorwstein) [inline deployment/ccip/changeset/deploy.go:496]: I'm not certain that rmnHome etc are threadsafe - maybe we come back to this idea in a follow up?

File: deployment/ccip/changeset/deploy_chain.go [hunk 27]
```diff
@@ -10,6 +10,11 @@ import (
 
 var _ deployment.ChangeSet[DeployChainContractsConfig] = DeployChainContracts
 
+// DeployChainContracts deploys all new CCIP v1.6 or later contracts for the given chains.
+// It returns the new addresses for the contracts.
+// If there is an error, it will return the successfully deployed addresses and the error so that the caller can call the
+// changeset again with the same input to retry the failed deployment.
```
Comment (connorwstein) [inline deployment/ccip/changeset/deploy_chain.go:16]: ah so it is already idempotent now? Lets explicitly mention that

File: deployment/ccip/changeset/initial_deploy_test.go (hunk 15 above)
Comment (connorwstein) [reply deployment/ccip/changeset/initial_deploy_test.go:16]: I see ok, we can come back to it then

File: deployment/environment.go [hunk 28]
```diff
@@ -76,6 +76,17 @@ type Environment struct {
 	Offchain          OffchainClient
 }
 
+func (e Environment) Copy() (Environment, error) {
```
Comment (connorwstein) [inline deployment/environment.go]: I think this is confusing because its only does a deep copy of one field. Its more clear to just use the NewEnvironment constructor in ApplyChangesets and remove this

File: deployment/common/changeset/test_helpers.go [hunk 29]
```diff
@@ -0,0 +1,87 @@
...
+	if err != nil {
+		return e, fmt.Errorf("failed to copy environment: %w", err)
+	}
+	for i, csa := range changesetApplications {
+		out, err := csa.Changeset(currentEnv, csa.Config)
+		if err != nil {
+			return e, fmt.Errorf("failed to apply changeset at index %d: %w", i, err)
+		}
+		if out.AddressBook != nil {
+			err := currentEnv.ExistingAddresses.Merge(out.AddressBook)
```
Comment (connorwstein) [inline deployment/common/changeset/test_helpers.go]: we're planning on making ExistingAddresses read-only as per CCIP-4247. So I'd invert this:
newAddresses = out.AddressBook.Merge(currentEnv.ExistingAddresses)

Comment (AnieeG) [reply deployment/common/changeset/test_helpers.go]: There is an issue with this. The consecutive changesets will load the state and try to look for previously deployed dependent contracts. that will fail if we do not update the existing addresses.

File: deployment/ccip/changeset/deploy.go (hunk 23 above)
Comment (AnieeG) [reply deployment/ccip/changeset/deploy.go:71]: It makes the DeployPrerequisite signature constant and helps adding more options. Otherwise for multicall there needs to be another bool addition to deployChainContracts

File: deployment/ccip/changeset/deploy.go (hunk 26 above)
Comment (AnieeG) [reply deployment/ccip/changeset/deploy.go:496]: This does not deal with any home contract write operation. Just deploys chain contracts in different chains

File: deployment/common/changeset/test_helpers.go (hunk 29 above)
Comment (AnieeG) [reply deployment/common/changeset/test_helpers.go]: Discussed offline and addressed

**Review by connorwstein — APPROVED** [review]

**Review by pramodhs2 — APPROVED** [review]

**Review by winder — APPROVED** [review]

**Review by kalverra — APPROVED** [review]

**Review by b-gopalswami — APPROVED** [review]
